src/palace/manager/core/metadata_layer.py

```diff
@@ -876,25 +876,22 @@ def __repr__(self) -> str:
 
         return description_string % description_data
 
-    def data_source(self, _db: Session) -> DataSource | None:
+    def data_source(self, _db: Session) -> DataSource:
         """Find the DataSource associated with this circulation information."""
         if not self.data_source_obj:
-            if self.data_source_name:
-                obj = DataSource.lookup(_db, self.data_source_name)
-                if not obj:
-                    raise ValueError("Data source %s not found!" % self._data_source)
-            else:
-                obj = None
+            obj = DataSource.lookup(_db, self.data_source_name)
+            if not obj:
+                raise ValueError("Data source %s not found!" % self._data_source)
             self.data_source_obj = obj
         return self.data_source_obj
 
-    def primary_identifier(self, _db: Session) -> Identifier | None:
+    def primary_identifier(self, _db: Session) -> Identifier:
         """Find the Identifier associated with this circulation information."""
         if not self.primary_identifier_obj:
             if self._primary_identifier:
                 obj, ignore = self._primary_identifier.load(_db)
             else:
-                obj = None
+                raise ValueError("No primary identifier provided!")
```

Comment:
In the calls `metadata_layer` makes to licensing, we are assuming these are not `None`. Since they don't seem to be `None` in our actual code, I modified the type hints and raise an `Exception` if they are None.